gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,434 @@
+from gtda.plotting import plot_point_cloud
+from gtda.local_homology import KNeighborsLocalVietorisRips, RadiusLocalVietorisRips
+from gtda.diagrams import PersistenceEntropy
+from sklearn.preprocessing import FunctionTransformer
+from gtda.pipeline import make_pipeline
+from dataclasses import dataclass
+
+
+def _local_homology_preprocess(
+    X: Tensor,
+    n_neighbors: Tuple[int, int],
+    homology_dimensions: Tuple[int, ...],
+    dim: int,
+) -> np.ndarray:
+    """
+    A helper file for topological regularizer
+    Computes the local homology, used for the linear surrogate function
+    Args:
+        X:
+            The point cloud on which the surrogate function is evaluated
+        n_neighbors:
+            The sizes of the KNN-neighborhoods to be considered on the first
+            and the second pass, respectively. An argument for
+            ``gtda.local_homology.KNeighborsLocalVietorisRips``
+        homology_dimensions:
+            what homology dimensions are computed
+        dim:
+            the homology dimension based on which the neighbors are computed
+            In the case of the topological regularizer, this is 1.
+    Returns:
+        conns:
+            An n by 2 numpy array describing the neighbors
+    """
+    kn_lh = KNeighborsLocalVietorisRips(
+        n_neighbors=n_neighbors, homology_dimensions=homology_dimensions
+    )
+    mod_pe = make_pipeline(
+        PersistenceEntropy(), FunctionTransformer(func=lambda X: 2 ** X)
+    )
+    pipe = make_pipeline(kn_lh, mod_pe)
+    loc_dim_features = pipe.fit_transform(X)
+    dimension = dim
+    dim_index = homology_dimensions.index(dimension)
+    colors = loc_dim_features[:, dim_index]
+    cutoffs = np.round(colors, 0)
+    nbrs = kn_lh.relevant_neighbors_.kneighbors()[1]
+    connections = []
+    for i in range(len(colors)):
+        cutoff = int(cutoffs[i])
+        trick = np.repeat(i, cutoff)
+        tmp_neighbors = nbrs[i]
+        tmp_trick = np.stack((nbrs[i][0:cutoff], trick))
+        maxes = np.amax(tmp_trick, 0)
+        mins = np.amin(tmp_trick, 0)
+        for j in range(len(maxes)):
+            connections.append([mins[j], maxes[j]])
+    conns = np.array(_unique_list(connections))
+    return conns
+
+
+def _create_dummy_loader(
+    tensor_x_t: Tensor,
+) -> DataLoader[Tuple[Tensor, ...]]:
+    """
+    A helper file for topological regularizer
+    A function for formatting the data for evaluating
+    the linear surrogate function.
+    Args:
+        tensor_x_t:
+            torch tensor representing the nodes of
+            the linear surrogate function.
+    Returns:
+        dummy_loader:
+            a torch dataloader based off of tensor_x_t
+    """
+    dummy = np.zeros(len(tensor_x_t))
+    dummy_ind = torch.from_numpy(dummy)
+    dummy_set = TensorDataset(tensor_x_t, dummy_ind)
+    dummy_loader = DataLoader(dummy_set, batch_size=4)
+    return dummy_loader
+
+
+def _unique_list(a_list: list) -> list:
```

Comment:
This can be much simplified, see my previous comment.

Reply:
yes, if hashable. Which is not the case here